weaver.go

```diff
@@ -131,35 +149,355 @@ func Run[T any, _ PointerToMain[T]](ctx context.Context, app func(context.Contex
 		return err
 	}
-	if wlet.info.RunMain {
-		main, err := wlet.GetImpl("weaver.Run", reflection.Type[T]())
-		if err != nil {
-			return err
+
+	for _, reg := range registrations {
+		if reg.Iface == reflection.Type[Main]() && wlet.Info.RunMain {
+			main, err := wlet.GetImpl("weaver.Run", reflection.Type[T]())
+			if err != nil {
+				return err
+			}
+			return app(ctx, main.(*T))
 		}
-		return app(ctx, main.(*T))
 	}
+
 	<-ctx.Done()
 	return ctx.Err()
 }
 
-func internalStart(ctx context.Context, opts private.AppOptions) (*weavelet, error) {
-	w, err := newWeavelet(ctx, opts, codegen.Registered())
-	if err != nil {
-		return nil, fmt.Errorf("error initializating application: %w", err)
-	}
-	if err := w.start(); err != nil {
-		return nil, err
-	}
-	return w, nil
+// Implements[T] is a type that is be embedded inside a component
+// implementation struct to indicate that the struct implements a component of
+// type T. For example, consider a Cache component.
+//
+//	type Cache interface {
+//	    Get(ctx context.Context, key string) (string, error)
+//	    Put(ctx context.Context, key, value string) error
+//	}
+//
+// A concrete type that implements the Cache component is written as follows:
+//
+//	type lruCache struct {
+//	    weaver.Implements[Cache]
+//	    ...
+//	}
+//
+// Because Implements is embedded inside the component implementation, methods
+// of Implements are available as methods of the component implementation type
+// and can be invoked directly. For example, given an instance c of type
+// lruCache, we can call c.Logger().
+type Implements[T any] struct {
+	// Component logger.
+	logger *slog.Logger
+
+	// Given a component implementation type, there is currently no nice way,
+	// using reflection, to get the corresponding component interface type [1].
+	// The component_interface_type field exists to make it possible.
+	//
+	// [1]: https://github.com/golang/go/issues/54393.
+	component_interface_type T //nolint:unused
+
+	// We embed implementsImpl so that component implementation structs
+	// implement the Unrouted interface by default but implement the
+	// RoutedBy[T] interface when they embed WithRouter[T].
+	implementsImpl
 }
 
-func init() {
-	// Provide weavertest with access to weavelet.
-	private.Start = func(ctx context.Context, opts private.AppOptions) (private.App, error) {
-		return internalStart(ctx, opts)
+// Logger returns a logger that associates its log entries with this component.
+func (i Implements[T]) Logger() *slog.Logger {
+	return i.logger
+}
+
+// implements is a method that can only be implemented inside the weaver
+// package. It exists so that a component struct that embeds Implements[T]
+// implements the InstanceOf[T] interface.
+//
+//nolint:unused
+func (Implements[T]) implements(T) {}
+
+// isImplements is an internal method that is only implemented by Implements[T]
+// and is used internally to check that a value is of type Implements[T].
+//
+//nolint:unused
```

Comment:
The GitHub Actions errors is fantastic!! :smile: 

`directive `//nolint:unused` is unused for linter "unused" (nolintlint)`

Reply:
:laughing: should be fixed now!